### PR TITLE
Remove the stat/expr distinction from `Transformers.Transformer`.

### PR DESCRIPTION
**Pull request by sjrd**
Recent changes have all but removed the distinction between statements and expressions in the IR. Notably:

* bb4f6da3628913639a82ad371c22731d9b2347f5 Allow Return arg to be a void if the target Labeled is a void.
* cdcff9973579ed77e87343f3edd9429661b4a786 Rename NoType to VoidType and print it as "void".

Therefore, it made no sense anymore that `Transformer.transform` had an `isStat` argument. In fact, the first of the two commits mentionned above semi-broke the semantics of that argument anyway. Moreover, `isStat` was never *used* in our codebase (only forwarded everywhere for no reason).

The only exception was a deserialization hack for IR <= 1.5. This one was already hacking around the fact the IR produced back then was not well-typed. See #4442 for context.

We now remove that parameter everywhere. It makes the deserialization hack a bit more verbose, but everything else becomes simpler.

### CODE REVIEW

**Comment by gzm0** [comment]
QQ: Did you verify that our basic IR backwards compatibility check catches issues with the deserialization hack? If no, how did you test the changes to it?

**Comment by sjrd** [comment]
No, our basic IR backward compat tests don't catch that. They can catch things that appear in our standard library, but not corner cases that only appear in the wild.

In fact I did not manage to reproduce a failure anymore with this hack completely disabled. I tried several things, including:

* Taking the report of https://github.com/scala-js/scala-js/issues/4601 with the library published with Scala.js 1.3.1 (but I think it did not cause an issue in the Scala 2-emitted IR, only the Scala 3-emitted one).
* Undo the compiler fixes of https://github.com/scala-js/scala-js/pull/4439/commits/e440de9b54fa0dda58cd274217800114546990dc to try and reproduce ill-typed IR today.

So yeah ... that's basically untested. 🤷‍♂️

**Comment by sjrd** [comment]
Ah, scratch that. I did manage to reproduce it! The following worked:

> * Undo the compiler fixes of [e440de9](https://github.com/scala-js/scala-js/commit/e440de9b54fa0dda58cd274217800114546990dc) to try and reproduce ill-typed IR today.

but I had to do it from a completely `clean` build. I had previously only re-compiled the test suite with those changes.

It turns out we have a "problematic" `If` here in our javalib:
https://github.com/scala-js/scala-js/blob/d3a5b0aed7434c6e9ac4e71ab01b8ea56e2a55a3/javalib/src/main/scala/java/util/regex/IndicesBuilder.scala#L500

So with the compiler fix undone, I get the IR checking error on this line. If I then unconditionally enable the hack (not just for IR <= 1.5), the IR checking error disappears and the tests pass (validating the hack on at least one input). I even tested on Wasm, which is much less forgiving of ill-typed IR even when it manages to go through our checkers.

File: project/BinaryIncompatibilities.scala [hunk 1]
```diff
@@ -10,6 +10,15 @@ object BinaryIncompatibilities {
     ProblemFilters.exclude[MissingClassProblem]("org.scalajs.ir.Trees$JSLinkingInfo$"),
     ProblemFilters.exclude[IncompatibleResultTypeProblem]("org.scalajs.ir.Trees#*.tpe"),
     ProblemFilters.exclude[MissingClassProblem]("org.scalajs.ir.Types$NoType$"),
+
+    // !!! Breaking, OK in minor release
```
Comment (gzm0) [inline project/BinaryIncompatibilities.scala:14]: Nit: Duplicate comment intended?

**Review by gzm0 — CHANGES_REQUESTED** [review]

File: project/BinaryIncompatibilities.scala (hunk 1 above)
Comment (sjrd) [reply project/BinaryIncompatibilities.scala:14]: Yes, that was intended. It's a different category of changes IMO.

File: ir/shared/src/main/scala/org/scalajs/ir/Transformers.scala [hunk 2]
```diff
@@ -19,204 +19,197 @@ import Version.Unversioned
...
-      trees.map(transformStat(_))
-
-    final def transformStat(tree: Tree): Tree =
-      transform(tree, isStat = true)
-
-    final def transformExpr(tree: Tree): Tree =
-      transform(tree, isStat = false)
-
-    def transformExprOrJSSpread(tree: TreeOrJSSpread): TreeOrJSSpread = {
+    final def transformTreeOrJSSpread(tree: TreeOrJSSpread): TreeOrJSSpread = {
```
[the remark this answers is not in the document]
Comment (sjrd) [reply ir/shared/src/main/scala/org/scalajs/ir/Transformers.scala:22]: I can't think of anything anyone could do with a spread other than keep is as a spread. 

Possibly one would want to transform the argument list, but for that you need to override `transform` anyway.

[the remark this answers is not in the document]
Comment (gzm0) [reply ir/shared/src/main/scala/org/scalajs/ir/Transformers.scala:22]: Fair enough. I can think of some extremely limited things, e.g. transforming `f(...[x])` to `f(x)` which would not need to look at the argument list as a whole.

Let's keep it final for now, we can always remove the final.
